File: python/vdm_service/modules/alarms.py
"""Alarm lifecycle ordering, deduplication and atomic notification decisions."""
import json
import time
from ..store import dumps, identity

TRANSITIONS = {"TRIGGERED","ESCALATED","DEESCALATED","SYNCED","RECOVERED","CANCELLED"}


def process(db, cid, did, event, max_age, max_pending_rows):
    event = dict(event)
    for key in ("eventId", "alarmId"):
        text = str(event.get(key, ""))
        if not text.isdecimal() or not 0 < int(text) < 2**64:
            raise ValueError(f"invalid {key}")
        event[key] = str(int(text))
    transition = str(event.get("transition", "")).removeprefix("ALARM_TRANSITION_")
    if transition not in TRANSITIONS:
        raise ValueError("invalid alarm transition")
    event["transition"] = transition
    if "level" in event:
        event["level"] = str(event["level"]).removeprefix("ALARM_LEVEL_")
    raw_ts = event.get("ts")
    if isinstance(raw_ts, bool) or not str(raw_ts).isdecimal():
        raise ValueError("alarm ts must be Unix seconds")
    ts = int(raw_ts)
    if ts > 2**63-1:
        raise ValueError("alarm ts too large")
    event["ts"] = ts
    now = time.time()
    if not db.execute("INSERT OR IGNORE INTO alarm_events VALUES(?,?,?,?,?,?)",
                      (cid,did,event["eventId"],event["alarmId"],dumps(event),now)).rowcount:
        return
    previous = db.execute("SELECT * FROM alarms WHERE connection_id=? AND device_id=? AND alarm_id=?",(cid,did,event["alarmId"])).fetchone()
    if previous and ts <= previous["ts"]:
        job_key = identity(cid,did,"reconcile")
        existing_job = db.execute('SELECT status FROM jobs WHERE key=?',(job_key,)).fetchone()
        if existing_job is None or existing_job['status'] == 'done':
            if db.execute("SELECT count(*) FROM jobs WHERE status<>'done'").fetchone()[0] >= max_pending_rows:
                raise BufferError('reconciliation job queue is full')
        db.execute("UPDATE alarms SET needs_reconcile=1 WHERE connection_id=? AND device_id=? AND alarm_id=?",(cid,did,event["alarmId"]))
        db.execute("INSERT INTO jobs(key,connection_id,device_id,created) VALUES(?,?,?,?) ON CONFLICT(key) DO UPDATE SET status='pending',attempts=0,due=0,created=excluded.created",(job_key,cid,did,now))
        return
    event["status"] = "CLOSED" if transition in {"RECOVERED","CANCELLED"} else "ACTIVE"
    event["currentActive"] = event["status"] == "ACTIVE"
    if event["status"] == "CLOSED":
        event.pop("level", None)
    # An unresolved ambiguity stays marked even when a newer event arrives.
    db.execute("INSERT INTO alarms VALUES(?,?,?,?,?,0) ON CONFLICT(connection_id,device_id,alarm_id) DO UPDATE SET ts=excluded.ts,payload=excluded.payload",
               (cid,did,event["alarmId"],ts,dumps(event)))
    prior = json.loads(previous['payload']) if previous else None
    levels = {'ALERT':1,'ALARM':2,'ACTION':3}
    previous_level = (prior.get('currentLevel', prior.get('level')) if prior else None)
    previous_rank = levels.get(str(previous_level).removeprefix('ALARM_LEVEL_'),0)
    current_rank = levels.get(event.get('level'),0)
    actionable = (transition in {'RECOVERED','CANCELLED'}
                  or transition == 'TRIGGERED' and previous is None
                  or transition == 'ESCALATED' and (previous is None or current_rank > previous_rank))
    if actionable and 0 <= now-ts <= max_age:
        if db.execute("SELECT count(*) FROM outbox WHERE status<>'done'").fetchone()[0] >= max_pending_rows:
            raise BufferError("notification outbox is full")
        key = identity(cid,did,event["eventId"],transition)
        payload = dict(connectionId=cid,deviceId=did,eventId=event["eventId"],transition=transition,alarm=event)
        db.execute("INSERT OR IGNORE INTO outbox(key,connection_id,device_id,payload,created) VALUES(?,?,?,?,?)",(key,cid,did,dumps(payload),now))


def reconcile(db, cid, did, response):
    body = response.get("getAlarmState", response.get("data", {}))
    if not isinstance(body, dict) or not isinstance(body.get("active", []), list):
        raise ValueError("invalid getAlarmState response")
    active = {str(item["alarmId"]): item for item in body.get("active", [])}
    for row in db.execute("SELECT * FROM alarms WHERE connection_id=? AND device_id=?",(cid,did)).fetchall():
        event = json.loads(row["payload"])
        item = active.pop(row["alarm_id"], None)
        event["status"] = "ACTIVE" if item else "CLOSED"
        event["currentActive"] = item is not None
        event["currentState"] = item
        event["reconciledAt"] = int(time.time())
        event["currentLevel"] = str(item["level"]).removeprefix("ALARM_LEVEL_") if item and "level" in item else None
        # Keep original eventId/transition/ts/level as historical evidence.
        # The device snapshot only changes the separate current lifecycle view.
        db.execute("UPDATE alarms SET payload=?,needs_reconcile=0 WHERE connection_id=? AND device_id=? AND alarm_id=?", (dumps(event),cid,did,row["alarm_id"]))
    for aid, item in active.items():
        event = {**item,"status":"ACTIVE","currentActive":True,"currentState":item,"reconciledAt":int(time.time()),"transition":"SYNCED","ts":0}
        if "level" in event:
            event["level"] = str(event["level"]).removeprefix("ALARM_LEVEL_")
        db.execute("INSERT INTO alarms VALUES(?,?,?,?,?,0)",(cid,did,aid,0,dumps(event)))

File: javascript/service/modules/alarms.js
"use strict";
const { register } = require("./register");
const { digest } = require("../identity");
// Lifecycle state, event deduplication and notification insertion share a transaction.
class AlarmStore {
  constructor(store) {
    this.db = store.db;
    this.options = store.options;
    this.job = store.job.bind(store);
  }
  alarms(c, d) {
    return this.db
      .prepare("SELECT * FROM alarms WHERE c=? AND d=?")
      .all(c, d)
      .map((r) => ({
        ...JSON.parse(r.payload),
        needsReconcile: Boolean(r.reconcile),
      }));
  }
  alarm(c, d, payload, now = Date.now() / 1000, maxAge = 300) {
    const e = {
      ...payload,
      transition: String(payload.transition || "").replace(
        "ALARM_TRANSITION_",
        "",
      ),
      level: payload.level
        ? String(payload.level).replace("ALARM_LEVEL_", "")
        : undefined,
    };
    for (const key of ["eventId", "alarmId"])
      if (
        !/^[1-9][0-9]*$/.test(String(e[key])) ||
        BigInt(e[key]) > 0xffffffffffffffffn
      )
        throw new Error(`invalid ${key}`);
    const terminal = ["RECOVERED", "CANCELLED"].includes(e.transition);
    if (
      ![
        "TRIGGERED",
        "ESCALATED",
        "DEESCALATED",
        "SYNCED",
        "RECOVERED",
        "CANCELLED",
      ].includes(e.transition) ||
      !Number.isFinite(Number(e.ts)) ||
      Number(e.ts) <= 0
    )
      throw new Error("invalid alarm");
    if (
      terminal
        ? e.level !== undefined
        : !["ALERT", "ALARM", "ACTION"].includes(e.level)
    )
      throw new Error("invalid alarm level");
    this.db.transaction(() => {
      const inserted = this.db
        .prepare("INSERT OR IGNORE INTO alarm_events VALUES(?,?,?,?,?,?,?)")
        .run(
          c,
          d,
          String(e.eventId),
          String(e.alarmId),
          Number(e.ts),
          JSON.stringify(e),
          now,
        ).changes;
      if (!inserted) return;
      const prev = this.db
        .prepare("SELECT * FROM alarms WHERE c=? AND d=? AND alarm=?")
        .get(c, d, String(e.alarmId));
      if (prev && Number(e.ts) <= prev.ts) {
        this.db
          .prepare(
            "UPDATE alarms SET reconcile=1 WHERE c=? AND d=? AND alarm=?",
          )
          .run(c, d, String(e.alarmId));
        this.job(c, d, "reconcile", {}, digest([c, d, "reconcile"]));
        return;
      }
      this.db
        .prepare(
          "INSERT INTO alarms(c,d,alarm,ts,payload) VALUES(?,?,?,?,?) ON CONFLICT(c,d,alarm) DO UPDATE SET ts=excluded.ts,payload=excluded.payload",
        )
        .run(c, d, String(e.alarmId), Number(e.ts), JSON.stringify(e));
      const old = prev ? JSON.parse(prev.payload) : null;
      const eligible =
        e.transition === "TRIGGERED"
          ? !old
          : e.transition === "ESCALATED"
            ? !old ||
              { ALERT: 1, ALARM: 2, ACTION: 3 }[e.level] >
                { ALERT: 1, ALARM: 2, ACTION: 3 }[old.level]
            : terminal;
      if (eligible && now - Number(e.ts) >= 0 && now - Number(e.ts) <= maxAge) {
        if (
          this.db
            .prepare("SELECT count(*) n FROM outbox WHERE status!='done'")
            .get().n >= this.options.maxInboxRows
        )
          throw new Error("notification outbox full");
        const id = digest([c, d, String(e.eventId), e.transition]);
        this.db
          .prepare(
            "INSERT OR IGNORE INTO outbox(id,c,d,payload,created) VALUES(?,?,?,?,?)",
          )
          .run(
            id,
            c,
            d,
            JSON.stringify({ connectionId: c, deviceId: d, event: e }),
            now,
          );
      }
    })();
  }
  reconcile(c, d, response) {
    const state = response.getAlarmState || response.data || response;
    if (state.active !== undefined && !Array.isArray(state.active))
      throw new Error("invalid alarm state reconciliation response");
    const active = new Map(
      (state.active || []).map((x) => [String(x.alarmId), x]),
    );
    this.db.transaction(() => {
      for (const row of this.db
        .prepare("SELECT * FROM alarms WHERE c=? AND d=? AND reconcile=1")
        .all(c, d)) {
        const payload = {
          ...JSON.parse(row.payload),
          currentActive: active.has(row.alarm),
          currentState: active.get(row.alarm) || null,
          reconciledAt: Date.now() / 1000,
        };
        this.db
          .prepare(
            "UPDATE alarms SET payload=?,reconcile=0 WHERE c=? AND d=? AND alarm=?",
          )
          .run(JSON.stringify(payload), c, d, row.alarm);
      }
    })();
  }
}
module.exports = { register, AlarmStore };
